discovery-provider/src/tasks/entity_manager/social_features.py

```diff
@@ -0,0 +1,175 @@
+from typing import Union
+
+from src.models.social.follow import Follow
+from src.models.social.repost import Repost
+from src.models.social.save import Save
+from src.tasks.entity_manager.utils import Action, EntityType, ManageEntityParameters
+
+action_to_record_type = {
+    Action.FOLLOW: EntityType.FOLLOW,
+    Action.UNFOLLOW: EntityType.FOLLOW,
+    Action.SAVE: EntityType.SAVE,
+    Action.UNSAVE: EntityType.SAVE,
+    Action.REPOST: EntityType.REPOST,
+    Action.UNREPOST: EntityType.REPOST,
+}
+
+create_actions = {Action.FOLLOW, Action.SAVE, Action.REPOST}
+delete_actions = {Action.UNFOLLOW, Action.UNSAVE, Action.UNREPOST}
+
+
+def create_social_record(params: ManageEntityParameters):
+
+    validate_social_feature(params)
+
+    create_record: Union[Save, Follow, Repost, None] = None
+    if params.action == Action.FOLLOW:
+        create_record = Follow(
+            blockhash=params.event_blockhash,
+            blocknumber=params.block_number,
+            created_at=params.block_datetime,
+            txhash=params.txhash,
+            follower_user_id=params.user_id,
+            followee_user_id=params.entity_id,
+            is_current=True,
+            is_delete=False,
+        )
+    if params.action == Action.SAVE:
+        create_record = Save(
+            blockhash=params.event_blockhash,
+            blocknumber=params.block_number,
+            created_at=params.block_datetime,
+            txhash=params.txhash,
+            user_id=params.user_id,
+            save_item_id=params.entity_id,
+            save_type=params.entity_type.lower(),
+            is_current=True,
+            is_delete=False,
+        )
+    if params.action == Action.REPOST:
+        create_record = Repost(
+            blockhash=params.event_blockhash,
+            blocknumber=params.block_number,
+            created_at=params.block_datetime,
+            txhash=params.txhash,
+            user_id=params.user_id,
+            repost_item_id=params.entity_id,
+            repost_type=params.entity_type.lower(),
+            is_current=True,
+            is_delete=False,
+        )
+
+    if create_record:
+        params.add_social_feature_record(
+            params.user_id,
+            params.entity_type,
+            params.entity_id,
+            action_to_record_type[params.action],
+            create_record,
+        )
+
+
+def delete_social_records(params):
+
+    validate_social_feature(params)
+
+    entity_key = (params.user_id, params.entity_type, params.entity_id)
+    record_type = action_to_record_type[params.action]
+    existing_entity = params.existing_records[record_type][entity_key]
```

Comment:
are we always sure the dict will have the keys here? afraid python will throw an error if we don't check for it. but maybe the flow already ensures that the keys exist in the dict?

Reply:
validation should cover this but i can add another check